client/rpc-api/src/state/mod.rs

```diff
@@ -136,4 +136,103 @@ pub trait StateApi<Hash> {
 	fn unsubscribe_storage(
 		&self, metadata: Option<Self::Metadata>, id: SubscriptionId
 	) -> RpcResult<bool>;
+
+	/// The `state_traceBlock` RPC provides a way to trace the re-execution of a single
+	/// block, collecting Spans and Events from both the client and the relevant WASM runtime.
+	/// The Spans and Events are conceptually equivalent to those from the [Tracing][1] crate.
+	///
+	/// The structure of the traces follows that of the block execution pipeline, so meaningful
+	/// interpretation of the traces requires an understanding of the Substrate chain's block
+	/// execution.
+	///
+	/// [Link to conceptual map of trace structure for Polkadot and Kusama block execution.][2]
+	///
+	/// [1]: https://crates.io/crates/tracing
+	/// [2]: https://docs.google.com/drawings/d/1vZoJo9jaXlz0LmrdTOgHck9_1LsfuQPRmTr-5g1tOis/edit?usp=sharing
+	///
+	/// ## Node requirements
+	///
+	/// - Fully synced archive node.
```

Comment:
What happens if you run this on a partially synced node, i.e. you call the RPC on a block that is present in the database (but we're still syncing). I would have expected that to work.
And the same question for non-archive nodes, again I would have expected the re-execution to work just as long as the block is actually available (i.e. hasn't been pruned).

Reply:
Just from trial:
-  a non-archive node seems to not keep any block bodies around so it fails quickly
- syncing nodes seem to have undefined behavior - most of the time it works but sometimes it doesn't. From my exprience of supporting api-sidecar I have seen borked data from queries to a actively syncing node

When I say syncing I mean a "major" sync, as in its trying to catch up on a large amount of blocks; not just actively syncing the latest blocks (which every node should be doing).

Reply:
Ok, good to know. A non-archive node should keep the most recent 256 blocks around though so that's a bit odd.